src/producer/TransactionMQProducer.cpp

```diff
@@ -0,0 +1,185 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "TransactionMQProducer.h"
+#include <string>
+#include "CommandHeader.h"
+#include "Logging.h"
+#include "MQClientFactory.h"
+#include "MQDecoder.h"
+#include "MessageSysFlag.h"
+#include "TransactionListener.h"
+#include "TransactionSendResult.h"
+
+using namespace std;
+namespace rocketmq {
+
+void TransactionMQProducer::initTransactionEnv() {}
+
+void TransactionMQProducer::destroyTransactionEnv() {}
+
+TransactionSendResult TransactionMQProducer::sendMessageInTransaction(MQMessage& msg, void* arg) {
+  if (nullptr == m_transactionListener) {
+    THROW_MQEXCEPTION(MQClientException, "transactionListener is null", -1);
+  }
+
+  SendResult sendResult;
+  msg.setProperty(MQMessage::PROPERTY_TRANSACTION_PREPARED, "true");
+  msg.setProperty(MQMessage::PROPERTY_PRODUCER_GROUP, getGroupName());
+  try {
+    sendResult = send(msg);
+  } catch (MQException& e) {
+    THROW_MQEXCEPTION(MQClientException, e.what(), -1);
+  }
+
+  LOG_DEBUG("sendMessageInTransaction result:%s", sendResult.toString().data());
+  LocalTransactionState localTransactionState = LocalTransactionState::UNKNOW;
+  switch (sendResult.getSendStatus()) {
+    case SendStatus::SEND_OK: {
+      try {
+        if (sendResult.getTransactionId() != "") {
+          msg.setProperty("__transactionId__", sendResult.getTransactionId());
+        }
+        string transactionId = msg.getProperty(MQMessage::PROPERTY_UNIQ_CLIENT_MESSAGE_ID_KEYIDX);
+        if (transactionId != "") {
+          msg.setTransactionId(transactionId);
+        }
+        LOG_DEBUG("sendMessageInTransaction, msgId:%s, transactionId:%s",
+                  sendResult.getMsgId().data(), transactionId.data());
+        localTransactionState = m_transactionListener->executeLocalTransaction(msg, arg);
+        if (localTransactionState != LocalTransactionState::COMMIT_MESSAGE) {
+          LOG_WARN("executeLocalTransaction ret not LocalTransactionState::commit, msg:%s", msg.toString().data());
+        }
+      } catch (MQException& e) {
+        THROW_MQEXCEPTION(MQClientException, e.what(), -1);
```

Comment:
all code in this block don't throw MQException. but need catch all exception, and don't throw it. tell it to broker as remark by endTransaction

Reply:
executeLocalTransaction may throw exception, if send prepare fail, no need send endTransaction to Broker

Reply:
Please see org.apache.rocketmq.client.impl.producer.DefaultMQProducerImpl#sendMessageInTransaction, java do it, but i don't understand.